File: Services/ApplicationAuthorization/TixFactory.ApplicationAuthorization/Operations/GetAuthorizedOperationsOperation.cs
﻿using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using TixFactory.ApplicationAuthorization.Entities;
using TixFactory.Operations;

namespace TixFactory.ApplicationAuthorization
{
	/// <summary>
	/// Retrieves names of operations authorized for a target application, for the given ApiKey.
	/// </summary>
	/// <remarks>
	/// As application (specified by ApiKey header, <see cref="GetAuthorizedOperationsRequest.TargetApplicationKey"/>),
	/// I have received a request with this ApiKey: <see cref="GetAuthorizedOperationsRequest.ApiKey"/>.
	/// Please tell me which operations the ApiKey I have received has access to.
	/// </remarks>
	internal class GetAuthorizedOperationsOperation : IAsyncOperation<GetAuthorizedOperationsRequest, ICollection<string>>
	{
		private readonly IApplicationEntityFactory _ApplicationEntityFactory;
		private readonly IApplicationKeyEntityFactory _ApplicationKeyEntityFactory;
		private readonly IApplicationKeyValidator _ApplicationKeyValidator;

		public GetAuthorizedOperationsOperation(
			IApplicationEntityFactory applicationEntityFactory,
			IApplicationKeyEntityFactory applicationKeyEntityFactory,
			IApplicationKeyValidator applicationKeyValidator)
		{
			_ApplicationEntityFactory = applicationEntityFactory ?? throw new ArgumentNullException(nameof(applicationEntityFactory));
			_ApplicationKeyEntityFactory = applicationKeyEntityFactory ?? throw new ArgumentNullException(nameof(applicationKeyEntityFactory));
			_ApplicationKeyValidator = applicationKeyValidator ?? throw new ArgumentNullException(nameof(applicationKeyValidator));
		}

		public async Task<(ICollection<string> output, OperationError error)> Execute(GetAuthorizedOperationsRequest request, CancellationToken cancellationToken)
		{
			var targetApplicationKey = await _ApplicationKeyEntityFactory.GetApplicationKey(request.TargetApplicationKey, cancellationToken).ConfigureAwait(false);
			if (targetApplicationKey == null || !targetApplicationKey.Enabled)
			{
				return (Array.Empty<string>(), null);
			}

			var targetApplication = _ApplicationEntityFactory.GetApplicationById(targetApplicationKey.ApplicationId);
			var authorizedOperations = await _ApplicationKeyValidator.GetAuthorizedOperations(targetApplication.Id, request.ApiKey, cancellationToken).ConfigureAwait(false);
			return (authorizedOperations, null);
		}
	}
}

File: Node/application-authorization-service/operations/GetAuthorizedOperationsOperation.js
export default class {
	constructor(logger, authorizationHandler, applicationEntityFactory, applicationKeyEntityFactory) {
		this.logger = logger;
		this.authorizationHandler = authorizationHandler;
		this.applicationEntityFactory = applicationEntityFactory;
		this.applicationKeyEntityFactory = applicationKeyEntityFactory;
	}

	get allowAnonymous() {
		return true;
	}

    get name() {
        return "GetAuthorizedOperations";
    }
	
	get requestParameters() {
		return ["apiKey"];
	}
 
    async execute(requestBody, request) {
		const targetApplicationKey = await this.applicationKeyEntityFactory.getApplicationKeyByGuid(request.apiKey);
		if (!targetApplicationKey) {
			return Promise.resolve([]);
		}

		const targetApplication = await this.applicationEntityFactory.getApplicationById(targetApplicationKey.applicationId);
		if (!targetApplication) {
			this.logger.warn("Valid ApiKey does not map to valid application.",
				"\n\tKey:", targetApplicationKey.name, `(${targetApplicationKey.id})`);

			return Promise.resolve([]);
		}

		const authorizedOperationNames = await this.authorizationHandler.loadAndCacheAuthorizedOperations(targetApplication.name, requestBody.apiKey);
		return Promise.resolve(authorizedOperationNames);
    }
};